### PR TITLE
Implement encode/decodeUri for Gpx provider

### PR DESCRIPTION
**Pull request by nyalldawson**
...and some other gpx provider cleanups

### CODE REVIEW

File: src/providers/gpx/qgsgpxfeatureiterator.cpp [hunk 1]
```diff
@@ -497,8 +497,8 @@ QgsGeometry *QgsGPXFeatureIterator::readTrackGeometry( const QgsTrack &trk )
 QgsGPXFeatureSource::QgsGPXFeatureSource( const QgsGPXProvider *p )
   : mFileName( p->mFileName )
   , mFeatureType( p->mFeatureType )
-  , indexToAttr( p->indexToAttr )
-  , mFields( p->attributeFields )
+  , indexToAttr( p->mIndexToAttr )
```
Comment (nirvn) [inline src/providers/gpx/qgsgpxfeatureiterator.cpp:500]: Not directly related to this PR, but a follow up cleanup that'd be nice is the renaming of QgsGPXFeatureSource private members indexToAttr and data to mIndexToAttr and mData too.

**Review by nirvn — APPROVED** [review]
Nice one.

File: src/providers/gpx/qgsgpxfeatureiterator.cpp (hunk 1 above)
Comment (nyalldawson) [reply src/providers/gpx/qgsgpxfeatureiterator.cpp]: done

Comment (nirvn) [reply src/providers/gpx/qgsgpxfeatureiterator.cpp]: @nyalldawson , thanks.

**Pull request merged**